### PR TITLE
[8.0] Add account constraint chronology

### PR DESCRIPTION
**Pull request by adrienpeiffer**
Add account_constraint_chronology module

This module help ensuring the chronology of invoice numbers.

It prevents the validation of invoices when:
- there are draft invoices with an anterior date
- there are validated invoices with a posterior date

The check can be activated on a per-journal basis
(for sale and purchase journals).

### CODE REVIEW

File: account_invoice_constraint_chronology/model/account.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+from openerp.osv import orm, fields
```
Comment (bwrsandman) [inline account_invoice_constraint_chronology/model/account.py]: This is a new module for 8.0.

Please make use of the new API like you do in account_invoice.py

**Comment by nbessi** [comment]
Please wait  pr #19 to be merged and ported to 8.0 before merging any new addons these way we will have good bases to start the 8.0 serie

**Comment by adrienpeiffer** [comment]
There is a problem with proforma invoice, i'm looking into it

**Comment by sbidoul** [comment]
:+1: code review and functional tests. Thanks!

File: account_invoice_constraint_chronology/model/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                                 ('date_invoice', '!=', False),
+                                 ('date_invoice', '<', inv.date_invoice),
+                                 ('journal_id', '=', inv.journal_id.id)])
+                if len(invoices) > 0:
+                    raise models.except_orm(_('Error !'),
+                                            _("Chronology Error!"
+                                              " Please confirm older draft"
+                                              " invoices before %s and"
+                                              " try again.") %
+                                            inv.date_invoice)
```
Comment (bwrsandman) [inline account_invoice_constraint_chronology/model/account_invoice.py]: You should format this according to the context. There is risk of date format mismatch

File: account_invoice_constraint_chronology/model/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                                             inv.date_invoice),
+                                            ('journal_id', '=',
+                                             inv.journal_id.id)])
+                    if len(invoices) > 0:
+                        raise models.except_orm(_('Error !'),
+                                                _("Chronology Error! There"
+                                                  " exist at least one invoice"
+                                                  " with a date posterior"
+                                                  " to %s.") %
+                                                inv.date_invoice)
```
Comment (bwrsandman) [inline account_invoice_constraint_chronology/model/account_invoice.py]: You should format this according to the context. There is risk of date format mismatch

File: account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    def test_invoice_validate(self):
+        journal_id = get_journal_check(self, True)
+        today = datetime.now()
+        tomorrow = today + timedelta(days=1)
+        date = tomorrow.strftime('%Y-%m-%d')
+        invoice = create_simple_invoice(self, journal_id, date)
+        workflow.trg_validate(self.uid, 'account.invoice', invoice.id,
+                              'invoice_open', self.cr)
+        date = today.strftime('%Y-%m-%d')
```
Comment (bwrsandman) [inline account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py]: Please use `DEFAULT_SERVER_DATE_FORMAT` from `openerp.tools`

File: account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+DB = common.DB
+ADMIN_USER_ID = common.ADMIN_USER_ID
+
+
+def get_simple_product_id(self):
+    return self.registry('product.product').create(self.cr,
+                                                   self.uid,
+                                                   {'name': 'product_test_01',
+                                                    'lst_price': 2000.00,
+                                                    }, context={})
```
Comment (bwrsandman) [inline account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py]: Context can be gotten with  `self.context = self.user_model.context_get(self.cr, self.uid)`

Comment (sbidoul) [inline account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py]: Does context matter in such a test?

Comment (bwrsandman) [inline account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py]: Tests can fail due to servers configured in different languages. `mail` does so.

File: account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py (hunk 4 above)
Comment (adrienpeiffer) [inline account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py]: :+1:

File: account_invoice_constraint_chronology/model/account_invoice.py (hunk 2 above)
Comment (adrienpeiffer) [inline account_invoice_constraint_chronology/model/account_invoice.py]: :+1:

File: account_invoice_constraint_chronology/model/account_invoice.py (hunk 3 above)
Comment (adrienpeiffer) [inline account_invoice_constraint_chronology/model/account_invoice.py]: :+1:

File: account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py (hunk 5 above)
Comment (adrienpeiffer) [inline account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py]: :+1:

File: account_invoice_constraint_chronology/model/account_invoice.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                                 ('date_invoice', '!=', False),
+                                 ('date_invoice', '<', inv.date_invoice),
+                                 ('journal_id', '=', inv.journal_id.id)])
+                if len(invoices) > 0:
+                    date_invoice_format = datetime\
+                        .strptime(inv.date_invoice,
+                                  DEFAULT_SERVER_DATE_FORMAT)
+                    date_invoice_tz = fields\
+                        .Date.context_today(self, date_invoice_format)
+                    raise models.except_orm(_('Error !'),
```
Comment (pedrobaeza) [inline account_invoice_constraint_chronology/model/account_invoice.py]: Better to import new `exceptions` package and use `exceptions.Warning`.

Comment (gurneyalex) [inline account_invoice_constraint_chronology/model/account_invoice.py]: No space before an exclamation mark in English, but you should remove the exclamation mark altogether as it is redundant with the icon displayed on screen.

File: account_invoice_constraint_chronology/model/account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                                            _("Chronology Error!"
```
Comment (gurneyalex) [inline account_invoice_constraint_chronology/model/account_invoice.py]: you already said it was an Error in the title of the dialog. Maybe "Chronology Error" could be used as title?

File: account_invoice_constraint_chronology/model/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                                            ('journal_id', '=',
+                                             inv.journal_id.id)])
+                    if len(invoices) > 0:
+                        date_invoice_format = datetime\
+                            .strptime(inv.date_invoice,
+                                      DEFAULT_SERVER_DATE_FORMAT)
+                        date_invoice_tz = fields\
+                            .Date.context_today(self, date_invoice_format)
+                        raise models.except_orm(_('Error !'),
+                                                _("Chronology Error! There"
```
Comment (gurneyalex) [inline account_invoice_constraint_chronology/model/account_invoice.py]: same 2 remarks as above

**Comment by gurneyalex** [comment]
could you also add a .pot file?

**Comment by bwrsandman** [comment]
:+1:

**Comment by sbidoul** [comment]
:+1:

File: account_invoice_constraint_chronology/model/account.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+from openerp import models, fields, api
+
+
+class account_jounrnal(models.Model):
```
Comment (nbessi) [inline account_invoice_constraint_chronology/model/account.py]: Typo jounrnal journal

File: account_invoice_constraint_chronology/view/account_view.xml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+        <record id="view_account_journal_form" model="ir.ui.view">
+            <field name="name">account.journal.form (account_constraint_chronology)</field>
+            <field name="model">account.journal</field>
+            <field name="inherit_id" ref="account.view_account_journal_form"/>
+            <field name="arch" type="xml">
+                <field name="entry_posted" position="after">
+                    <field name="check_chronology" attrs="{'readonly': [('type', 'not in', ['sale', 'purchase', 'sale_refund', 'purchase_refund'])]}"/>
+                </field>
+                <xpath expr="//field[@name='type']" position="attributes">
+                    <attribute name="on_change">on_change_type(type)</attribute>
```
Comment (nbessi) [inline account_invoice_constraint_chronology/view/account_view.xml]: Seems wrong you should simply use the api.onchange decorator if it is not inserted in view it is a bug

File: account_invoice_constraint_chronology/model/account.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    _inherit = ['account.journal']
+
+    check_chronology = fields.Boolean(string='Check Chronology', default=False)
+
+    @api.multi
```
Comment (nbessi) [inline account_invoice_constraint_chronology/model/account.py]: use api.onchange decorator

File: account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+from openerp import exceptions
+from datetime import datetime, timedelta
+from openerp.tools import DEFAULT_SERVER_DATE_FORMAT
+
+DB = common.DB
+ADMIN_USER_ID = common.ADMIN_USER_ID
+
+
+def get_simple_product_id(self):
+    return self.registry('product.product').create(self.cr,
```
Comment (nbessi) [inline account_invoice_constraint_chronology/tests/test_account_constraint_chronology.py]: please use self.env['model'] registery is deprecated

File: account_invoice_constraint_chronology/view/account_view.xml [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+                </xpath>
+             </field>
+        </record>
+    </data>
+</openerp>
```
Comment (nbessi) [inline account_invoice_constraint_chronology/view/account_view.xml]: Missing white space at end of file

**Comment by nbessi** [comment]
:-1: see my comments in line

**Comment by nbessi** [comment]
:+1:

**Comment by sbidoul** [comment]
@gurneyalex the needfixing label can be removed from this one

**Comment by gurneyalex** [comment]
@adrienpeiffer Hello, since this is a new addon, could you send back a signed CLA to the association? See http://odoo-community.org/page/website.cla for the required paperwork. 

cc @sbidoul 

Thanks

**Comment by sbidoul** [comment]
@gurneyalex I sent an Entity CLA for Acsone. Does it not cover contributions of Acsone employees?

**Comment by gurneyalex** [comment]
On 09/09/2014 16:59, Stéphane Bidoul (ACSONE) wrote:

> @gurneyalex https://github.com/gurneyalex I sent an Entity CLA for
> Acsone. Does it not cover contributions of Acsone employees?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/account-financial-tools/pull/21#issuecomment-54982004.
> 
> hmm my bad...

## 

Alexandre Fayolle
Chef de Projet
Tel : + 33 (0)4 79 26 57 94

Camptocamp France SAS
Savoie Technolac, BP 352
73377 Le Bourget du Lac Cedex
http://www.camptocamp.com

**Comment by gurneyalex** [comment]
:+1:

File: account_invoice_constraint_chronology/model/account_invoice.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        res = super(account_invoice, self).action_move_create()
+        for inv in self:
+            if inv.journal_id.check_chronology:
+                invoices = \
+                    self.search([('state', 'not in',
+                                  ['open', 'paid', 'cancel', 'proforma',
+                                   'proforma2']),
+                                 ('date_invoice', '!=', False),
+                                 ('date_invoice', '<', inv.date_invoice),
+                                 ('journal_id', '=', inv.journal_id.id)])
```
Comment (sbidoul) [inline account_invoice_constraint_chronology/model/account_invoice.py]: Adding limit=1 may help with performance

File: account_invoice_constraint_chronology/model/account_invoice.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                                               " invoices before %s and"
+                                               " try again.") %
+                                             date_invoice_tz)
+
+                if inv.internal_number is False:
+                    invoices = self.search([('state', 'in', ['open', 'paid']),
+                                            ('date_invoice', '>',
+                                             inv.date_invoice),
+                                            ('journal_id', '=',
+                                             inv.journal_id.id)])
```
Comment (sbidoul) [inline account_invoice_constraint_chronology/model/account_invoice.py]: Adding limit=1 may help with performance

**Comment by sbidoul** [comment]
:+1:

**Comment by sbidoul** [comment]
This one should be good for merging, now.

**Pull request merged**